src/VisualStudio/IntegrationTest/TestUtilities/InProcess/SolutionExplorer_InProc.cs

```diff
@@ -431,17 +431,27 @@ private static void WaitForDesignMode()
             // state believing a debugger session was active.
             //
             // This delay should be replaced with a proper wait condition once the correct one is determined.
-            var editAndContinueService = GetComponentModelService<IEditAndContinueService>();
-            do
+            var debugService = GetComponentModelService<VisualStudioWorkspace>().Services.GetRequiredService<IDebuggingWorkspaceService>();
+            using (var debugSessionEndedEvent = new ManualResetEventSlim(initialState: false))
             {
-                if (stopwatch.Elapsed >= Helper.HangMitigatingTimeout)
+                debugService.BeforeDebuggingStateChanged += (_, e) =>
                 {
-                    throw new TimeoutException("Failed to enter design mode in a timely manner.");
+                    if (e.After == DebuggingState.Design)
+                    {
+                        debugSessionEndedEvent.Set();
+                    }
+                };
+
+                if (dte.Debugger.CurrentMode == EnvDTE.dbgDebugMode.dbgDesignMode)
```

Comment:
❔ Is there a way to do this check without DTE?

Reply:
Not that I know of. IDebuggingWorkspaceService can use some redesign, but that's outside of the scope of the PR.